### ISSUE TITLE
Scalability

### ISSUE THREAD

**Issue by jcbrand** [collective/collective.zws#1]
Hi @datakurre 

This add-on actually looks very promising to me. It would be great to have websocket support in Plone. Are there any major challenges and why do you think it might not scale in a production environment?

Cheers
JC

**Comment by datakurre** [comment]
It should scale, if websockets are handled by dedicated zeo clients.

Otherwise many open web sockets may affect the http server performance. It all depends on the asyncore loop performance: https://hg.python.org/cpython/file/2.7/Lib/asyncore.py#l125

Other issue, of course, is that the used websocket library is based on outdated spec https://github.com/datakurre/collective.zws/blob/master/src/collective/zws/websocket.py

**Comment by jcbrand** [comment]
Thanks.

I started to think about websockets and Zope/Plone since we're going to use websockets to show generated document previews as they become available in PloneIntranet.

The backend there is a Tornado server but it made me wonder how feasible it would be to use websockets in Plone.

As a fun project I would consider extending XMPP chat in Plone to work over websockets.

**Comment by datakurre** [comment]
@jcbrand Feel free to continue on `collective.zws`. For me, it never got beyond being just experiment, and I never found more recent websocket server code for it.

Instead, we have been using RabbitMQ with webstomp (browser connects to RabbitMQ using websockets and Plone listens RabbitMQ). Authentication, of course, could be handled more robust with direct websocket connection. Some old demo at https://www.youtube.com/watch?v=lDfKNEDgJB8

**Comment by petri** [comment]
As far as I know, websocket connections are in typical use cases fairly long-lasting, in comparison to plain-vanilla HTTP requests. Wouldn't that mean that the Zope/ZEO thread pool would run out almost immediately in any system with more than just a few users? One would have to have quite a few ZEO clients with many threads each, making the associated (RAM etc.) costs prohibitively high, I would assume. 

Perhaps that could be remedied by designing applications so that they only use a websocket connection when it's really needed and/or adds extra value. But I wonder if that defeats the purpose (of using websockets to begin with) to a degree that the whole thing no longer makes sense. I don't know.

**Comment by datakurre** [comment]
@petri Not really. WebSocket connections are managed asynchronously by asyncore main thread and only individual messages are handled in worker threads.

The master branch only included request / response style websocket usage, which did not really offer anything beyond technological demo. The another branch used ZMQ to broadcast events from worker threads to to active websocket connections managed by the main asyncore thread.

Yet, this is dead in water as long as there isn't maintained websocket server implementation for asyncore. And that is not likely to be happen, because asyncore cannot handle large number of connections (it loops through all connections to poll for possible socket traffic).
